presto-main/src/main/java/com/facebook/presto/operator/ScanFilterAndProjectOperator.java

```diff
@@ -173,7 +176,15 @@ public final boolean isFinished()
             finishing = true;
         }
 
-        return finishing && pageBuilder.isEmpty();
+        if (!finishing || !pageBuilder.isEmpty()) {
+            return false;
+        }
+
+        operatorContext.getSystemMemoryTracker().freeMemory(recordedPageBuilderMemoryUsage);
```

Comment:
I don't think this is the right place to be freeing system memory. `isFinished()` indicates that the operator is done, but this doesn't mean it's actually going to be garbage collected. I would probably put this in `Driver.destroyIfNecessary` where the other memory is released